noir_stdlib/src/meta/expr.nr

```diff
@@ -1,131 +1,209 @@
+
 use crate::option::Option;
 use crate::meta::op::UnaryOp;
 use crate::meta::op::BinaryOp;
 
 impl Expr {
+    /// If this expression is an array literal `[elem1, ..., elemN]`, this returns a slice of each element in the array.
     #[builtin(expr_as_array)]
     // docs:start:as_array
     fn as_array(self) -> Option<[Expr]> {}
     // docs:end:as_array
 
+    /// If this expression is an assert, this returns the assert expression and the optional message.
     #[builtin(expr_as_assert)]
     // docs:start:as_assert
     fn as_assert(self) -> Option<(Expr, Option<Expr>)> {}
     // docs:end:as_assert
 
+    /// If this expression is an assert_eq, this returns the left-hand-side and right-hand-side
+    /// expressions, together with the optional message.
     #[builtin(expr_as_assert_eq)]
     // docs:start:as_assert_eq
     fn as_assert_eq(self) -> Option<(Expr, Expr, Option<Expr>)> {}
     // docs:end:as_assert_eq
 
+    /// If this expression is an assignment, this returns a tuple with the left hand side
+    /// and right hand side in order.
     #[builtin(expr_as_assign)]
     // docs:start:as_assign
     fn as_assign(self) -> Option<(Expr, Expr)> {}
     // docs:end:as_assign
 
-    #[builtin(expr_as_integer)]
-    // docs:start:as_integer
-    fn as_integer(self) -> Option<(Field, bool)> {}
-    // docs:end:as_integer
-
+    /// If this expression is a binary operator operation `<lhs> <op> <rhs>`,
+    /// return the left-hand side, operator, and the right-hand side of the operation.
     #[builtin(expr_as_binary_op)]
     // docs:start:as_binary_op
     fn as_binary_op(self) -> Option<(Expr, BinaryOp, Expr)> {}
     // docs:end:as_binary_op
 
+    /// If this expression is a block `{ stmt1; stmt2; ...; stmtN }`, return
+    /// a slice containing each statement.
     #[builtin(expr_as_block)]
     // docs:start:as_block
     fn as_block(self) -> Option<[Expr]> {}
     // docs:end:as_block
 
+    /// If this expression is a boolean literal, return that literal.
     #[builtin(expr_as_bool)]
     // docs:start:as_bool
     fn as_bool(self) -> Option<bool> {}
     // docs:end:as_bool
 
+    /// If this expression is a cast expression `expr as type`, returns the casted
+    /// expression and the type to cast to.
+    // docs:start:as_cast
     #[builtin(expr_as_cast)]
     fn as_cast(self) -> Option<(Expr, UnresolvedType)> {}
+    // docs:end:as_cast
 
+    /// If this expression is a `comptime { stmt1; stmt2; ...; stmtN }` block,
+    /// return each statement in the block.
     #[builtin(expr_as_comptime)]
     // docs:start:as_comptime
     fn as_comptime(self) -> Option<[Expr]> {}
     // docs:end:as_comptime
 
+    /// If this expression is a constructor `Type { field1: expr1, ..., fieldN: exprN }`,
+    /// return the type and the fields.
+    #[builtin(expr_as_constructor)]
+    // docs:start:as_constructor
+    fn as_constructor(self) -> Option<(Quoted, [(Quoted, Expr)])> {}
```

Comment:
I think we can return the Path as a UnresolvedType in this case.

Would probably have to add some support for re-parsing those when we're working on parsing a constructor but I think it would be clearer from a user's perspective than just a Quoted value.

Reply:
> And then, if we unquote it, when we parse it as a constructor we can't know if that UnresolvedType is actually a Named one or not.

This sounds fine to me, we could check during name resolution assuming it is interned.

I'm debating changing the constructor parser itself to match as well. Maybe it should always parse a type then later on we just confirm it is a struct type.

Reply:
Done!